keps/sig-node/4960-container-stop-signals/README.md

```diff
@@ -0,0 +1,382 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Container runtimes let you define a STOPSIGNAL to let your container images change which signal is delivered to kill the container. Currently you can only configure this by defining STOPSIGNAL in the container image definition file before you build the image. This becomes difficult to change when you’re using prebuilt images. This KEP proposes to add support to configure custom stop signals for containers from the ContainerSpec. Kubernetes has no equivalent for STOPSIGNAL as part of Pod or Container APIs. This KEP proposes to add support to configure custom stop signals for containers from the ContainerSpec.
+
+## Motivation
+
+Container runtimes like Docker lets you configure the stop signal with which a container would be killed when you start a container. This can be configured either from the container image definition file itself with the STOPSIGNAL instruction or by using the `--stop-signal` flag when starting a container with the respective CLI tool for your runtime. Currently there is no equivalent to this in the Kubernetes APIs.
+
+While managing containers with Kubernetes, if you want to customize an existing image by changing its predefined stop signal or override the default stop signal of SIGTERM, currently you would have to rebuild the container image and update the stop signal at the image definition level. 
+
+Having stop signal as a first class citizen in the container specification would make it easier for users to set custom stop signals for their containers across all types of workloads.
+
+### Goals
+
+- Add a new StopSignal field to ContainerSpec
+- Update the CRI API to take StopSignal as a field in the StopContainerRequest and pass it down to the container runtime via the StopContainer method of remoteRuntimeService
+- Update StopContainer implementation in container runtimes to use the container’s StopSignal, if defined as the stop signal when killing containers.
+
+### Non-Goals
+
+- Change any existing behaviour with how stop signals work when defined in the container image
+
+## Proposal
+
+A new StopSignal field will be added to the Container spec. This new field can take string values, and can be used to define a stop signal when creating Pods.
+
+```go
+// pkg/apis/core/types.go
+type Container struct {
+  // ...
+  // +optional
+  StopSignal string
+}
+
+type ContainerStatus struct {
+  // ...
+  // +optional
+  StopSignal string
+}
+```
+
+The CRI API would be updated to have StopSignal as the third field in `StopContainerRequest`, along with `container_id` and `timeout`. 
```

Comment:
I would like to see what we are actually adding to the CRI API here. 

So my experience with adding new CRI APIs fields is that it is difficult to add both the CRI API and the container runtime in the same release. What I have done in the past is define the CRI API and roll that out as a first alpha. And then you can implement that functionality in the container runtime once containerd/CRI-O have been updated to consume the first alpha. Otherwise you hit a nasty dependency chain where you need the CRI published before you can use those fields in the container runtime.